File: exo3.6.py
#!/usr/bin/env python3
import os
import sys
import tempfile

# Programme pour calculer C(n,k) sans utiliser de boucles
# Utilise des fichiers temporaires et la récursion via fork()

def main():
    if len(sys.argv) != 3:
        sys.exit(0)
    
    n = int(sys.argv[1])
    k = int(sys.argv[2])
    
    if n < k:
        return 1
    
    if k == 0 or k == n:
        print("1")
        return 0
    
    # Création de fichiers temporaires
    with tempfile.NamedTemporaryFile(mode='w+', delete=False) as f1:
        file1 = f1.name
    
    with tempfile.NamedTemporaryFile(mode='w+', delete=False) as f2:
        file2 = f2.name
    
    pid1 = os.fork()
    if pid1 == 0:
        # Premier fils: calcule C(n-1, k)
        # Rediriger stdout vers file1
        with open(file1, 'w') as f:
            sys.stdout = f
            os.execl(sys.argv[0], sys.argv[0], str(n-1), str(k))
        sys.exit(0)
    
    pid2 = os.fork()
    if pid2 == 0:
        # Deuxième fils: calcule C(n-1, k-1)
        with open(file2, 'w') as f:
            sys.stdout = f
            os.execl(sys.argv[0], sys.argv[0], str(n-1), str(k-1))
        sys.exit(0)
    
    # Père attend ses deux fils
    os.wait()
    os.wait()
    
    if pid1 != 0 and pid2 != 0:
        # Lecture des résultats
        with open(file1, 'r') as f1:
            res1 = int(f1.read().strip())
        
        with open(file2, 'r') as f2:
            res2 = int(f2.read().strip())
        
        print(f"Resultat : {res1 + res2}")
        
        # Nettoyage des fichiers temporaires
        os.unlink(file1)
        os.unlink(file2)
    
    return 0

if __name__ == "__main__":
    main()

File: exo3.6.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>


/*Ecrire un programme pour calculerCnk qui n'utilise aucune boucle (ni while ni for), et qui n'ait
comme seule fonction que la fonction main. La fonction main ne doit contenir aucun appel a
elle-meme.On pourra utiliser des fichiers textes temporaires dans le repertoire /tmp.*/
int main (int argc, char** argv) {

    if (argc != 3) {
        exit (0);
    }

    int n = atoi (argv[1]);
    int k = atoi (argv[2]);

    if (n < k) {
        return 1;
    }

    if (k == 0 || k == n) {
        printf ("1\n");
        return 0;
    }

    char file1[64], file2[64];

    sprintf (file1, "/home/fabien/S3/tmp_res_%d_%d_a.txt", n, k);
    sprintf (file2, "/home/fabien/S3/tmp_res_%d_%d_b.txt", n, k);

    pid_t pid1 = fork ();
    if (pid1 == 0) {
        char sn[16], sk[16];
        
        snprintf (sn, sizeof (sn), "%d", n - 1);
        snprintf (sk, sizeof (sk), "%d", k);

        freopen (file1, "w", stdout);
        execl (argv[0], argv[0], sn, sk, NULL);

        exit (0);
    }

    pid_t pid2 = fork ();
    if (pid2 == 0) {
        char sn[16], sk[16];

        snprintf (sn, sizeof (sn), "%d", n - 1);
        snprintf (sk, sizeof (sk), "%d", k - 1);

        freopen (file2, "w", stdout);
        execl (argv[0], argv[0], sn, sk, NULL);

        exit (0);
    }

    wait (NULL);
    wait (NULL);

    if (pid1 != 0 && pid2 != 0) {
        int res1, res2;
    
        FILE* f1 = fopen (file1, "r");
        FILE* f2 = fopen (file2, "r");

        fscanf (f1, "%d", &res1);
        fscanf (f2, "%d", &res2);

        fclose (f1); fclose (f2);

        printf ("Resultat : %d\n", res1 + res2);

        unlink (file1);
        unlink (file2);
    }
    return 0;
}